randgen: prevent geo shape initialization

msbutler opened the pull request:
This patch prevents allocating the heavy geography shapes at initializion,
which reduces the runtime of server startup under stress from 2.1 to 1.7
seconds on average. Running the following cmd:
```
/dev test pkg/server -f TestServerStartup/SystemTenantOnly --stress --count 100 --ignore-cache
```
With this patch:
```
Stats over 100 runs: max = 2.0s, min = 0.9s, avg = 1.7s, dev = 0.2s
```
Without this patch:
```
Stats over 100 runs: max = 2.5s, min = 1.2s, avg = 2.1s, dev = 0.2s
```

Note that the new `BenchmarkTestServerStartup` does not capture the benefits of
this change, as it benchmarks performance after initialization.

Epic: none

Release note: none

msbutler committed:
randgen: prevent geo shape initialization

This patch prevents allocating the heavy geography shapes at initializion,
which reduces the runtime of server startup under stress from 2.1 to 1.7
seconds on average. Running the following cmd:
```
/dev test pkg/server -f TestServerStartup/SystemTenantOnly --stress --count 100 --ignore-cache
```
With this patch:
```
Stats over 100 runs: max = 2.0s, min = 0.9s, avg = 1.7s, dev = 0.2s
```
Without this patch:
```
Stats over 100 runs: max = 2.5s, min = 1.2s, avg = 2.1s, dev = 0.2s
```

Note that the new `BenchmarkTestServerStartup` does not capture the benefits of
this change, as it benchmarks measure performance after initialization.

Epic: none

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142040)
<!-- Reviewable:end -->

msbutler committed:
server: add TestServer startup tests

While #142039 added a test server startup benchmark, this patch adds a vanilla
test server startup test which can be used to measure startup performance under
stress and the effects of changes to initialization.

Epic: none

Release note: none

msbutler commented:
Before:
<img width="1728" alt="image" src="https://github.com/user-attachments/assets/23640b37-9dc8-4cb9-9c18-7dbe37049ec7" />

After:
<img width="1727" alt="image" src="https://github.com/user-attachments/assets/85ff656e-0c54-4144-8436-c72c1bfcb519" />

rafiss commented on pkg/server/testserver_test.go:
```diff
@@ -40,18 +35,53 @@ func BenchmarkTestServerStartup(b *testing.B) {
 					},
-				})
+				}
+				// Disable leader fortification as it currently causes large variability in runtime.
```
is there a way we can force the test to always run with `COCKROACH_INTERNAL_DISABLE_METAMORPHIC_TESTING` set? or something else to make metamorphic constants not get used?

https://github.com/cockroachdb/cockroach/blob/4e70581d869daf562c6433b59ac9a3eb5eabcd25/pkg/util/metamorphic/constants.go#L270-L285

(this seems generally useful for all microbenchmarks):

rafiss reviewed (approved):
this is great!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142040#-:-OK7SQ2m2oSSKYQr9PLh:bqk7sj1)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @mgartner, @msbutler, and @srosenberg)

___
*[`pkg/sql/randgen/datum.go` line 603 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142040#-OK7SETr7KT6OKcQ1OVi:-OK7SETr7KT6OKcQ1OVj:b-54drse) ([raw file](https://github.com/cockroachdb/cockroach/blob/85016597bc12073248a1cf02f3e7bb9babce6032/pkg/sql/randgen/datum.go#L603)):*
> ```Go
> var randInterestingDatums map[types.Family][]tree.Datum
> 
> func getRandInterestingDatums(typ types.Family) ([]tree.Datum, bool) {
> ```

nit: a comment saying why we did this would be helpful. and maybe mention that the function behaves the same as if you were accessing a map.


<!-- Sent from Reviewable.io -->

msbutler reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142040)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @mgartner, and @srosenberg)

___
*[`pkg/sql/randgen/datum.go` line 603 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142040#-OK7SETr7KT6OKcQ1OVi:-OK7X0EN-LexE-yui0a-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/85016597bc12073248a1cf02f3e7bb9babce6032/pkg/sql/randgen/datum.go#L603)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: a comment saying why we did this would be helpful. and maybe mention that the function behaves the same as if you were accessing a map.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner commented:
Nice fix!
